File: sdk/typescript/src/protocol.ts
import { IncompatibleProtocolError, ProtocolError } from "./errors.js";
import type { AgentEvent, ApprovalResponse, ErrorDetail, VerificationReport } from "./types.js";

export const MAX_FRAME_BYTES = 1_048_576;
export type RecordValue = Record<string, unknown>;
export interface Hello {
  type: "hello";
  protocol_version: number;
  server_version: string;
  max_frame_bytes: number;
  capabilities: string[];
}
export type Frame = Hello
  | { type: "response"; id: string | null; result?: RecordValue; error?: ErrorDetail }
  | { type: "event"; request_id: string | null; session_id: string; event: AgentEvent };

export function object(value: unknown): value is RecordValue {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function string(value: unknown): value is string { return typeof value === "string"; }
function boolean(value: unknown): value is boolean { return typeof value === "boolean"; }
function integer(value: unknown): value is number { return typeof value === "number" && Number.isInteger(value); }
function unsigned(value: unknown): value is number { return integer(value) && value >= 0; }
function strings(value: unknown): value is string[] { return Array.isArray(value) && value.every(string); }
function nullable(value: unknown, check: (v: unknown) => boolean): boolean { return value === null || check(value); }
function detail(value: unknown): value is ErrorDetail {
  return object(value) && string(value.code) && string(value.message);
}
function blobs(value: unknown): boolean {
  return Array.isArray(value) && value.every(v => object(v) && string(v.sha256) && unsigned(v.bytes));
}
function operation(value: unknown): boolean {
  if (!object(value)) return false;
  switch (value.type) {
    case "read": return strings(value.paths) && boolean(value.external);
    case "write": return strings(value.paths) && boolean(value.external) && boolean(value.destructive);
    case "bash": return string(value.command) && string(value.cwd) && unsigned(value.timeout_ms)
      && ["read_only", "mutating", "unknown"].includes(String(value.class));
    case "web_search": return string(value.query) && boolean(value.contains_workspace_data);
    case "web_open": return string(value.url) && boolean(value.private_target);
    default: return false;
  }
}
export function verificationReport(value: unknown): value is VerificationReport {
  return object(value) && string(value.name) && string(value.command) && string(value.cwd)
    && unsigned(value.timeout_ms)
    && ["not_run", "running", "passed", "failed", "cancelled", "denied", "interrupted"].includes(String(value.status))
    && nullable(value.operation_id, string) && nullable(value.started_at_ms, unsigned)
    && nullable(value.finished_at_ms, unsigned) && nullable(value.exit_code, integer)
    && blobs(value.output_refs) && nullable(value.message, string);
}
function event(value: unknown): value is AgentEvent {
  if (!object(value)) return false;
  switch (value.type) {
    case "text": return string(value.text);
    case "status": return string(value.message);
    case "approval": return object(value.request) && string(value.request.operation_id)
      && string(value.request.summary) && operation(value.request.operation) && "arguments" in value.request;
    case "tool_started": return string(value.operation_id) && string(value.name) && string(value.context);
    case "tool_output": return string(value.call_id) && string(value.stream) && string(value.chunk);
    case "tool_completed": return string(value.operation_id) && object(value.result)
      && string(value.result.call_id) && string(value.result.output) && boolean(value.result.is_error)
      && boolean(value.result.truncated) && blobs(value.result.blob_refs) && "metadata" in value.result;
    case "agent_updated": {
      const v = value.snapshot;
      return object(v) && string(v.id) && string(v.role) && string(v.objective) && string(v.profile)
        && ["queued", "running", "completed", "failed", "cancelled"].includes(String(v.state))
        && nullable(v.phase, string) && nullable(v.active_operation, string)
        && strings(v.changed_files) && nullable(v.last_error, string);
    }
    case "usage": return object(value.usage) && unsigned(value.usage.input_tokens)
      && unsigned(value.usage.output_tokens) && unsigned(value.usage.cached_input_tokens);
    case "verification": return verificationReport(value.report);
    case "runtime": return "event" in value;
    case "done": return ["completed", "cancelled", "failed"].includes(String(value.status))
      && (value.error === undefined || detail(value.error));
    default: return false;
  }
}
export function approvalResponse(value: unknown): value is ApprovalResponse {
  return value === "approve_once" || value === "approve_session" || value === "deny"
    || (object(value) && Object.keys(value).length === 1 && object(value.edit)
      && Object.keys(value.edit).length === 1 && "arguments" in value.edit);
}
export function parseFrame(bytes: Uint8Array): Frame {
  let value: unknown;
  try { value = JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes)); }
  catch { throw new ProtocolError(); }
  if (!object(value)) throw new ProtocolError();
  if (value.type === "hello") {
    if (value.protocol_version !== 1) throw new IncompatibleProtocolError();
    const capabilities = value.capabilities;
    if (!string(value.server_version) || value.max_frame_bytes !== MAX_FRAME_BYTES || !strings(capabilities)
      || !["prompt", "stream", "approval", "cancel", "resume", "verify"].every(c => capabilities.includes(c))) {
      throw new IncompatibleProtocolError();
    }
    return value as unknown as Hello;
  }
  if (value.type === "response" && (value.id === null || (string(value.id) && /^[1-9][0-9]*$/.test(value.id)))) {
    if (detail(value.error) && !("result" in value)) return value as unknown as Frame;
    if (object(value.result) && !("error" in value) && value.id !== null) return value as unknown as Frame;
  }
  if (value.type === "event" && (value.request_id === null || (string(value.request_id) && /^[1-9][0-9]*$/.test(value.request_id)))
    && string(value.session_id) && event(value.event)) return value as unknown as Frame;
  throw new ProtocolError();
}

/** One fixed-size buffer avoids quadratic copying for arbitrarily fragmented frames. */
export class FrameDecoder {
  private readonly buffer = Buffer.allocUnsafe(MAX_FRAME_BYTES);
  private length = 0;
  constructor(private readonly accept: (frame: Frame, bytes: number) => void) {}
  push(chunk: Buffer): void {
    let offset = 0;
    while (offset < chunk.length) {
      const lf = chunk.indexOf(10, offset);
      const end = lf < 0 ? chunk.length : lf;
      const count = end - offset;
      if (this.length + count > MAX_FRAME_BYTES) throw new ProtocolError("The Kurama process exceeded the 1 MiB frame limit.");
      if (this.length === 0 && lf >= 0) {
        this.accept(parseFrame(chunk.subarray(offset, end)), count);
      } else {
        chunk.copy(this.buffer, this.length, offset, end);
        this.length += count;
        if (lf >= 0) {
          const size = this.length;
          this.length = 0;
          this.accept(parseFrame(this.buffer.subarray(0, size)), size);
        }
      }
      if (lf < 0) break;
      offset = lf + 1;
    }
  }
  end(): void {
    if (this.length !== 0) throw new ProtocolError("The Kurama process closed stdout in the middle of a frame.");
  }
}

File: sdk/python/src/kurama/_protocol.py
"""Protocol v1 decoding. Invalid input never appears in diagnostic messages."""

from __future__ import annotations

import json
from typing import Any, cast

from .errors import ProtocolError
from .types import (
    AgentSnapshot,
    AgentUpdatedEvent,
    ApprovalEvent,
    ApprovalRequest,
    BlobRef,
    DoneEvent,
    ErrorInfo,
    Event,
    Operation,
    RuntimeEvent,
    StatusEvent,
    TextEvent,
    ToolCompletedEvent,
    ToolOutputEvent,
    ToolResult,
    ToolStartedEvent,
    Usage,
    UsageEvent,
    VerificationEvent,
    VerificationReport,
)

PROTOCOL_VERSION = 1
MAX_FRAME_BYTES = 1_048_576
CAPABILITIES = {"prompt", "stream", "approval", "cancel", "resume", "verify"}


def object_value(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ProtocolError("Expected a protocol object.")
    return value


def string(value: Any) -> str:
    if not isinstance(value, str):
        raise ProtocolError("Expected a protocol string.")
    return value


def integer(value: Any) -> int:
    if type(value) is not int:
        raise ProtocolError("Expected a protocol integer.")
    return value


def boolean(value: Any) -> bool:
    if type(value) is not bool:
        raise ProtocolError("Expected a protocol boolean.")
    return value


def array(value: Any) -> list[Any]:
    if not isinstance(value, list):
        raise ProtocolError("Expected a protocol array.")
    return value


def optional_string(value: Any) -> str | None:
    return None if value is None else string(value)


def optional_integer(value: Any) -> int | None:
    return None if value is None else integer(value)


def enum(value: Any, choices: set[str]) -> Any:
    if not isinstance(value, str) or value not in choices:
        raise ProtocolError("Unrecognized protocol variant.")
    return value


def request_id(value: Any) -> str:
    identifier = string(value)
    if (
        not identifier
        or not identifier.isascii()
        or not identifier.isdecimal()
        or identifier[0] == "0"
    ):
        raise ProtocolError("Invalid response/request correlation ID.")
    return identifier


def _reject_constant(value: str) -> None:
    raise ValueError("Non-finite JSON number")


def decode_frame(frame: bytes) -> dict[str, Any]:
    if not frame.endswith(b"\n"):
        raise ProtocolError("The process ended with a truncated protocol frame.")
    if len(frame) - 1 > MAX_FRAME_BYTES:
        raise ProtocolError("Protocol frame exceeds the 1048576-byte limit.")
    try:
        value = json.loads(frame[:-1].decode("utf-8"), parse_constant=_reject_constant)
    except (UnicodeDecodeError, ValueError, RecursionError):
        raise ProtocolError("The process sent invalid UTF-8 JSON.") from None
    return object_value(value)


def decode_error(value: Any) -> ErrorInfo:
    value = object_value(value)
    return ErrorInfo(string(value.get("code")), string(value.get("message")))


def decode_blob(value: Any) -> BlobRef:
    value = object_value(value)
    return BlobRef(string(value.get("sha256")), integer(value.get("bytes")))


def decode_report(value: Any) -> VerificationReport:
    value = object_value(value)
    return VerificationReport(
        name=string(value.get("name")),
        command=string(value.get("command")),
        cwd=string(value.get("cwd")),
        timeout_ms=integer(value.get("timeout_ms")),
        status=enum(
            value.get("status"),
            {
                "not_run",
                "running",
                "passed",
                "failed",
                "cancelled",
                "denied",
                "interrupted",
            },
        ),
        operation_id=optional_string(value.get("operation_id")),
        started_at_ms=optional_integer(value.get("started_at_ms")),
        finished_at_ms=optional_integer(value.get("finished_at_ms")),
        exit_code=optional_integer(value.get("exit_code")),
        output_refs=[decode_blob(item) for item in array(value.get("output_refs"))],
        message=optional_string(value.get("message")),
    )


def decode_operation(value: Any) -> Operation:
    value = object_value(value)
    kind = enum(value.get("type"), {"read", "write", "bash", "web_search", "web_open"})
    if kind in {"read", "write"}:
        for path in array(value.get("paths")):
            string(path)
        boolean(value.get("external"))
        if kind == "write":
            boolean(value.get("destructive"))
    elif kind == "bash":
        string(value.get("command"))
        string(value.get("cwd"))
        enum(value.get("class"), {"read_only", "mutating", "unknown"})
        integer(value.get("timeout_ms"))
    elif kind == "web_search":
        string(value.get("query"))
        boolean(value.get("contains_workspace_data"))
    elif kind == "web_open":
        string(value.get("url"))
        boolean(value.get("private_target"))
    return cast(Operation, value)


def decode_event(value: Any) -> Event:
    value = object_value(value)
    kind = value.get("type")
    if kind == "text":
        return TextEvent(string(value.get("text")))
    if kind == "approval":
        request = object_value(value.get("request"))
        return ApprovalEvent(
            ApprovalRequest(
                string(request.get("operation_id")),
                decode_operation(request.get("operation")),
                string(request.get("summary")),
                request.get("arguments", {}),
            )
        )
    if kind == "tool_started":
        return ToolStartedEvent(
            string(value.get("operation_id")),
            string(value.get("name")),
            string(value.get("context")),
        )
    if kind == "tool_output":
        return ToolOutputEvent(
            string(value.get("call_id")),
            string(value.get("stream")),
            string(value.get("chunk")),
        )
    if kind == "tool_completed":
        result = object_value(value.get("result"))
        return ToolCompletedEvent(
            string(value.get("operation_id")),
            ToolResult(
                string(result.get("call_id")),
                string(result.get("output")),
                boolean(result.get("is_error")),
                result.get("metadata"),
                boolean(result.get("truncated")),
                [decode_blob(item) for item in array(result.get("blob_refs"))],
            ),
        )
    if kind == "agent_updated":
        snapshot = object_value(value.get("snapshot"))
        return AgentUpdatedEvent(
            AgentSnapshot(
                string(snapshot.get("id")),
                string(snapshot.get("role")),
                string(snapshot.get("objective")),
                string(snapshot.get("profile")),
                enum(
                    snapshot.get("state"),
                    {"queued", "running", "completed", "failed", "cancelled"},
                ),
                optional_string(snapshot.get("phase")),
                optional_string(snapshot.get("active_operation")),
                [string(item) for item in array(snapshot.get("changed_files"))],
                optional_string(snapshot.get("last_error")),
            )
        )
    if kind == "usage":
        usage = object_value(value.get("usage"))
        return UsageEvent(
            Usage(
                integer(usage.get("input_tokens")),
                integer(usage.get("output_tokens")),
                integer(usage.get("cached_input_tokens")),
            )
        )
    if kind == "status":
        return StatusEvent(string(value.get("message")))
    if kind == "verification":
        return VerificationEvent(decode_report(value.get("report")))
    if kind == "runtime":
        return RuntimeEvent(object_value(value.get("event")))
    if kind == "done":
        status = enum(value.get("status"), {"completed", "cancelled", "failed"})
        return DoneEvent(
            status, decode_error(value["error"]) if "error" in value else None
        )
    raise ProtocolError(
        "Unrecognized friendly event type; install compatible SDK and Kurama versions."
    )
